Refs #28643 -- Added NullIf database function.

atombrella opened the pull request.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,25 @@ will result in a database error.
+
+.. versionadded:: 2.1
+
+Accepts two arguments, and return ``NULL`` if they are equal, otherwise
```
return -> returns
NULL or None? (add an example for that case?)
otherwise returns expression1

timgraham commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class NullIfTests(TestCase):
+    def setUp(self):
```
Did you consider using setUpTestData?

timgraham commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,36 @@
+    def test_nullif(self):
+        authors = Author.objects.annotate(nullif=NullIf('alias', 'name')).values_list('nullif')
+
+        self.assertSequenceEqual(
```
A single line looks okay?

timgraham commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,36 @@
+
+    def test_nullif(self):
+        authors = Author.objects.annotate(nullif=NullIf('alias', 'name')).values_list('nullif')
+
```
Omit blank lines.

timgraham commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,36 @@
+    def test_too_few_args(self):
+        msg = "'NullIf' takes exactly 2 arguments (1 given)"
+        with self.assertRaisesMessage(TypeError, msg):
+            list(Author.objects.annotate(nullif=NullIf('name')))
```
Isn't `NullIf('name')` enough to raise the exception?

timgraham commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,36 @@
+        self.assertSequenceEqual(
+            authors, [
+                ('smithj',),
+                (None,),
```
On Oracle, this returns an empty string instead of None.

atombrella replied:
OK. That will essentially require two separate tests, one for Oracle, and one for the other backends?

atombrella commented:
buildbot, test on oracle.

pope1ni commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,33 @@
+        Author.objects.create(name='John Smith', alias='smithj')
+        Author.objects.create(name='Rhonda', alias='Rhonda')
+
+    @skipIf(connection.vendor == 'Oracle', "Oracle doesn't use NULL, but the empty string")
```
What is it that causes Oracle to return an empty string?

According to their [documentation](https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions102.htm) this shouldn't be the case:

> If they are equal, then the function returns null.

It sounds like there is either a problem in Django or cx_Oracle?

If this is going to remain like this, either we need a workaround for Oracle to force `NULL` or it should be documented.

felixxm replied:
@pope1ni It's a convention rather than an issue, please see [#null-and-empty-strings](https://docs.djangoproject.com/en/dev/ref/databases/#null-and-empty-strings) documentation and this [code](https://github.com/django/django/blob/master/django/db/backends/oracle/operations.py#L159-L167). If we will use fields with types other than `BinaryField` and `CharField` then `NullIf` should return `None`, e.g:
```python
from decimal import Decimal

from django.db.models.functions import NullIf
from django.test import TestCase

from .models import DecimalModel

class NullIfTests(TestCase):

    @classmethod
    def setUpTestData(cls):
        DecimalModel.objects.create(n1=Decimal('1.1'), n2=Decimal('1.3'))
        DecimalModel.objects.create(n1=Decimal('1.2'), n2=Decimal('1.2'))

    def test_nullif(self):
        decimals = DecimalModel.objects.annotate(nullif=NullIf('n1', 'n2')).values_list('nullif')
        self.assertSequenceEqual(decimals, [(Decimal('1.1'),), (None,)])
```

pope1ni replied:
I see. Wasn't aware of that quirk as I don't use Oracle - thanks for the explanation.

I think in that case, a test with another type such as the example you provided should definitely be added, and the documentation for `NullIf` should highlight this still substitution applies to it's output - after all, one typically expects the value from this function to be `None` or the first argument.

atombrella replied:
I'll put a link to that section in the documentation in this patch.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

atombrella commented on django/db/models/functions/comparison.py:
```diff
@@ -87,3 +87,9 @@ def __init__(self, *expressions, **extra):
+
+class NullIf(Func):
+    function = 'NULLIF'
+    name = 'NullIf'
```
I think those `name` attributes should also be removed in `django.db.models.functions.window`.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

pope1ni commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,39 @@ will result in a database error.
+
+    Please note that due to conventions, this functions returns the empty
+    string instead of ``NULL``. The :ref:`Oracle notes <oracle-notes>` discuss
+    this.
```
```Due to convention, this function returns empty string instead of ``None``. See the :ref:`Oracle notes <oracle-notes>` for more details.```

atombrella commented:
buildbot, test on oracle.

pope1ni commented on docs/releases/2.1.txt:
```diff
@@ -169,6 +169,9 @@ Models
   it in model forms.
 
+* The new database function :class:`~django.db.models.functions.NullIf`
+  provides the NULLIF function.
```
```The new :class:`~django.db.models.functions.NullIf` database function returns `None` if the two provided expressions are equal.```

atombrella replied:
I think it should be double ticks around `None`, but otherwise I copied it in. Thanks.

pope1ni commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,46 @@
+        with self.assertRaisesMessage(TypeError, msg):
+            NullIf('name')
+
+    @skipUnless(connection.vendor == 'oracle', 'oracle specific test for NULL-literal')
```
`oracle specific...` → `Oracle specific...`

pope1ni commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,46 @@
+        authors = Author.objects.annotate(nullif=NullIf('alias', 'name')).values_list('nullif')
+        self.assertSequenceEqual(authors, [('smithj',), (None,)])
+
+    @skipUnless(connection.vendor == 'oracle', 'oracle specific test for the NULL return value')
```
`oracle specific...` → `Oracle specific...`

pope1ni commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,46 @@
+        Author.objects.create(name='John Smith', alias='smithj', age=40)
+        Author.objects.create(name='Rhonda', alias='Rhonda', age=45)
+
+    @skipIf(connection.vendor == 'oracle', 'oracle uses the empty string instead')
```
`oracle uses...` → `Oracle uses...`

pope1ni commented on django/db/models/functions/comparison.py:
```diff
@@ -87,3 +87,14 @@ def __init__(self, *expressions, **extra):
+    def as_oracle(self, compiler, connection):
+        if (isinstance(self.get_source_expressions()[0], Value) and
+                self.get_source_expressions()[0].value is None):
+            raise ValueError('Oracle does not allow the literal NULL for expression1.')
```
Avoid calling `self.get_source_expressions()` twice and it becomes clearer:

```python
...
expression1 = self.get_source_expressions()[0]
if isinstance(expression1, Value) and expression1.value is None:
    raise ValueError('Oracle does not allow Value(None) for expression1.')
...
```

I've also tweaked the exception message to make it based on the argument provided in Python.

pope1ni commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,39 @@ will result in a database error.
+    >>> from django.db.models.functions import NullIf
+    >>> Author.objects.create(name='Margaret Smith', goes_by='Maggie')
+    >>> author = Author.objects.annotate(
+    ...    screen_name=NullIf('goes_by', 'name')).get()
```
Hanging indent please.

pope1ni commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,39 @@ will result in a database error.
+    Maggie
+
+    >>> author = Author.objects.annotate(
+    ...    screen_name=NullIf('goes_by', Value('Maggie'))).get()
```
Hanging indent please.

felixxm commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,46 @@
+    @skipUnless(connection.vendor == 'oracle', 'Test another type because empty string is treated as NULL')
+    def test_nullif_oracle_int(self):
+        authors = Author.objects.annotate(nullif=NullIf('age', Value(40))).values_list('nullif')
+        self.assertSequenceEqual(authors, [('',), (45,)])
```
You don't need to add a Oracle-specific test for types other than `BinaryField` and `CharField`, it will work exactly the same as for other back-ends. Please see my [comment](https://github.com/django/django/pull/9543#discussion_r160059507).

atombrella replied:
For this reason, I added a new field to the `Author` model.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
buildbot, test on oracle

atombrella commented:
buildbot, test oracle.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
@felixxm Maybe you can trigger the oracle build?

timgraham commented:
Special permissions aren't required. "buildbot, test on oracle." should work.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
@timgraham I tried earlier today with that command, and nothing happened. I see that it has started now.

atombrella commented:
buildbot, test on oracle.

atombrella commented:
I can't really figure out to get `test_nullif_oracle_binaryfield` to pass.

felixxm commented:
Oracle translates `NULLIF` into `CASE WHEN expression1 = expression2 THEN NULL ELSE expression1 END` (see [documentation](https://docs.oracle.com/database/121/SQLRF/functions128.htm#SQLRF00681)) and you cannot compare binary fields with `=` operator on Oracle. IMO you can remove `test_nullif_oracle_binaryfield`, because probably everyone who works with Oracle is aware of restrictions related with `BLOB`s. You cannot use them in most cases.

felixxm commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,39 @@ will result in a database error.
+Accepts two arguments, and returns ``None`` if they are equal, otherwise
+returns ``expression1``.
+
+.. note:: Caveats when using Oracle
```
I would use `.. admonition::` instead of `.. note::`.

felixxm commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,39 @@ will result in a database error.
+
+.. note:: Caveats when using Oracle
+
+    Due to convention, this function returns empty string instead of ``None``.
```
I think we should mention that this applies only to `CharField`s.

pope1ni replied:
Surely `TextField` also?

felixxm replied:
`IFNULL` will not work with `TextField` (`NCLOB`) from the same reason as for `BinaryField` (`BLOB`) 😒, see [documentation](https://docs.oracle.com/database/121/ADLOB/adlob_sql_semantics.htm#GUID-D8F66A2A-4D17-49C3-ADB2-BE384510DD6D).

felixxm commented on tests/db_functions/test_nullif.py:
```diff
@@ -0,0 +1,49 @@
+    @skipUnless(connection.vendor == 'oracle', 'Oracle specific test for the NULL return value')
+    def test_nullif_oracle_charfield(self):
+        authors = Author.objects.annotate(nullif=NullIf('alias', 'name')).values_list('nullif')
+        self.assertSequenceEqual(authors, [('smithj',), ('',)])
```
I think you can combine `test_nullif` and `test_nullif_oracle_charfield` into single test and use `connection.features.interprets_empty_strings_as_nulls` to determine `empty_value` (see [example](https://github.com/django/django/blob/master/tests/model_forms/tests.py#L278)).

atombrella commented:
buildbot, test on oracle.

atombrella commented:
I updated the commit message to credit reviewers.

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,41 @@ will result in a database error.
+
+.. versionadded:: 2.1
+
+Accepts two arguments, and returns ``None`` if they are equal, otherwise
```
arguments -> expressions?

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,41 @@ will result in a database error.
+    Due to convention, this function returns the empty string instead of
+    ``None`` when the expressions are of type
+    :class:`~django.db.models.CharField`. See the :ref:`Oracle notes
+    <oracle-notes>` for more details.
```
Link to the "NULL and empty strings" section instead?
chop "for more details"

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,41 @@ will result in a database error.
+    :class:`~django.db.models.CharField`. See the :ref:`Oracle notes
+    <oracle-notes>` for more details.
+
+    It is not possible to pass ``Value(None)`` to ``expression1`` as Oracle
```
Passing `Value(None)`` to ``expression1`` is prohibited on Oracle since Oracle doesn't accept...

timgraham commented on docs/ref/models/database-functions.txt:
```diff
@@ -149,6 +149,41 @@ will result in a database error.
+    >>> author = Author.objects.annotate(screen_name=NullIf('goes_by', 'name')).get()
+    >>> print(author.screen_name)
+    Maggie
+
```
Maybe add comments like some of the other examples? That being said, these examples doesn't seem particularly useful/realistic. Did you consider anything else?
```
>>> # screen_name will be None for any Authors with a goes_by equal to their name, or 'name' otherwise.
...
>>> # screen_name will be None for any Authors with a goes_by of 'Maggie', or 'goes_by' otherwise.
```

atombrella replied:
I think this was mostly copy pasting from previous examples. `NullIf` is good for a nice trick that I read about, guarding against division by zero (`10 / nullif(0, 0)` gives `null`). And together with `Coalesce` it could also be used. I'll try to rework them.

pope1ni commented:
@atombrella Are you going to be able to continue carrying this forward? It doesn't feel like there is much left to do....

felixxm commented:
@atombrella Do you have time to finish this patch?

atombrella commented:
I've rebased the patch, and updated version numbers, but the example still needs to be revised. I can finish it, as it's only documentation that needs revision.

atombrella commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

atombrella commented:
@pope1ni Please and thanks! :+1: I think the only missing thing is improving the documentation with a proper example. From what I get the function is commonly used together with `COALESCE`.

I've used my time on another project and work, so the two PRs that I currently have open have been neglected.
